GG/src/Layout.cpp

```diff
@@ -433,21 +437,37 @@ void Layout::SizeMove(const Pt& ul, const Pt& lr)
               Y(m_row_params[it->second.first_row].current_origin));
         Pt lr(X(m_column_params[it->second.last_column - 1].current_origin + m_column_params[it->second.last_column - 1].current_width),
               Y(m_row_params[it->second.last_row - 1].current_origin + m_row_params[it->second.last_row - 1].current_width));
+        Pt ul_margin(X0, Y0);
+        Pt lr_margin(X0, Y0);
         if (0 < it->second.first_row)
-            ul.y += static_cast<int>(m_cell_margin / 2);
+            ul_margin.y += static_cast<int>(m_cell_margin / 2);
         if (0 < it->second.first_column)
-            ul.x += static_cast<int>(m_cell_margin / 2);
+            ul_margin.x += static_cast<int>(m_cell_margin / 2);
         if (it->second.last_row < m_row_params.size())
-            lr.y -= static_cast<int>(m_cell_margin / 2.0 + 0.5);
+            lr_margin.y += static_cast<int>(m_cell_margin / 2.0 + 0.5);
         if (it->second.last_column < m_column_params.size())
-            lr.x -= static_cast<int>(m_cell_margin / 2.0 + 0.5);
+            lr_margin.x += static_cast<int>(m_cell_margin / 2.0 + 0.5);
+
+        ul += ul_margin;
+        lr -= lr_margin;
 
         if (it->second.alignment == ALIGN_NONE) { // expand to fill available space
             it->first->SizeMove(ul, lr);
         } else { // align as appropriate
             Pt available_space = lr - ul;
             Pt min_usable_size = it->first->MinUsableSize();
             Pt min_size = it->first->MinSize();
+
+            // HACK! This is put here so that TextControl, which is currently GG's
```

Comment:
The special case for `TextControl` in the above line 247 predates this PR.

It is necessary here because `TextControl` does not have a well defined minimum size.  This makes sure that the text is re-flows to the width of this column before asking the minimum size.  Otherwise t adds padding to the wrong size of text box.

You are right.  There should probably be a category of Wnd with scalable sizes that Layout can correctly handle.